File: microservices/shared/cache/redis-manager.js
class RedisManager {
  constructor(redisClient) {
    this.client = redisClient;
  }

  async set(key, value, ttl = 3600) {
    await this.client.setEx(key, ttl, JSON.stringify(value));
  }

  async get(key) {
    const value = await this.client.get(key);
    return value ? JSON.parse(value) : null;
  }

  async del(key) {
    await this.client.del(key);
  }

  async publish(channel, message) {
    await this.client.publish(channel, JSON.stringify(message));
  }
}

module.exports = RedisManager;

File: microservices/python-services/shared/cache/redis_manager.py
import json
from shared.database.connections import db

class RedisManager:
    @staticmethod
    async def set(key, value, ttl=3600):
        await db.redis_client.setex(key, ttl, json.dumps(value, default=str))

    @staticmethod
    async def get(key):
        value = await db.redis_client.get(key)
        return json.loads(value) if value else None

    @staticmethod
    async def delete(key):
        await db.redis_client.delete(key)

    @staticmethod
    async def publish(channel, message):
        await db.redis_client.publish(channel, json.dumps(message, default=str))

    @staticmethod
    async def cache_market_data(symbol, exchange, data, ttl=60):
        key = f"market:{symbol}:{exchange}"
        await RedisManager.set(key, data, ttl)

    @staticmethod
    async def cache_user_session(user_id, session_data, ttl=86400):
        key = f"session:{user_id}"
        await RedisManager.set(key, session_data, ttl)

    @staticmethod
    async def cache_portfolio(user_id, portfolio_data, ttl=1800):
        key = f"portfolio:{user_id}"
        await RedisManager.set(key, portfolio_data, ttl)